File: api/src/adapters/memory-graph.ts
/**
 * LLM-Memory-Graph Adapter
 *
 * Thin adapter for consuming context lineage and graph-based interaction
 * metadata from LLM-Memory-Graph.
 */

import { logger } from '../logger';
import {
  AdapterHealth,
  EcosystemAdapter,
  TimeRange,
  healthyAdapter,
  unhealthyAdapter,
} from './types';

export interface MemoryGraphConfig {
  endpoint: string;
  apiKey?: string;
  timeoutMs?: number;
}

/**
 * Context lineage
 */
export interface ContextLineage {
  lineageId: string;
  rootContextId: string;
  createdAt: Date;
  depth: number;
  nodes: LineageNode[];
  edges: LineageEdge[];
  metadata: LineageMetadata;
}

export interface LineageNode {
  nodeId: string;
  nodeType: 'user_message' | 'assistant_response' | 'system_prompt' | 'tool_call' | 'tool_result' | 'context_injection' | 'summary';
  createdAt: Date;
  contentHash: string;
  tokenCount: number;
  attributes: Record<string, unknown>;
}

export interface LineageEdge {
  edgeId: string;
  sourceNodeId: string;
  targetNodeId: string;
  edgeType: 'follows' | 'references' | 'summarizes' | 'derived_from' | 'tool_invocation';
  weight: number;
  createdAt: Date;
}

export interface LineageMetadata {
  totalTokens: number;
  totalInteractions: number;
  activeBranches: number;
  compressionRatio: number;
}

/**
 * Interaction graph
 */
export interface InteractionGraph {
  graphId: string;
  sessionId: string;
  createdAt: Date;
  lastUpdated: Date;
  statistics: GraphStatistics;
  topics: TopicCluster[];
  entities: EntityReference[];
}

export interface GraphStatistics {
  nodeCount: number;
  edgeCount: number;
  avgDegree: number;
  clusteringCoefficient: number;
  diameter: number;
  density: number;
}

export interface TopicCluster {
  clusterId: string;
  topic: string;
  relevanceScore: number;
  nodeIds: string[];
  keywords: string[];
}

export interface EntityReference {
  entityId: string;
  entityType: string;
  name: string;
  firstMentioned: Date;
  mentionCount: number;
  nodeIds: string[];
}

/**
 * Memory snapshot
 */
export interface MemorySnapshot {
  snapshotId: string;
  sessionId: string;
  createdAt: Date;
  contextWindowTokens: number;
  summarizedTokens: number;
  activeMemories: ActiveMemory[];
  retrievalStats: RetrievalStats;
}

export interface ActiveMemory {
  memoryId: string;
  memoryType: 'short_term' | 'long_term' | 'working' | 'episodic' | 'semantic';
  contentPreview: string;
  tokenCount: number;
  relevanceScore: number;
  lastAccessed: Date;
  accessCount: number;
}

export interface RetrievalStats {
  totalRetrievals: number;
  avgLatencyMs: number;
  cacheHitRate: number;
  relevanceAvg: number;
}

/**
 * Graph analytics
 */
export interface GraphAnalytics {
  periodStart: Date;
  periodEnd: Date;
  totalSessions: number;
  totalNodesCreated: number;
  totalEdgesCreated: number;
  avgSessionDepth: number;
  avgSessionTokens: number;
  topTopics: string[];
  memoryEfficiency: MemoryEfficiency;
}

export interface MemoryEfficiency {
  avgCompressionRatio: number;
  cacheHitRate: number;
  retrievalLatencyP50Ms: number;
  retrievalLatencyP99Ms: number;
}

/**
 * Query parameters for lineage
 */
export interface LineageQuery {
  contextId?: string;
  sessionId?: string;
  startTime?: Date;
  endTime?: Date;
  minDepth?: number;
  maxDepth?: number;
  includeContent?: boolean;
}

/**
 * LLM-Memory-Graph adapter for consuming graph data
 */
export class MemoryGraphAdapter implements EcosystemAdapter {
  private config: MemoryGraphConfig;
  private connected: boolean = false;

  constructor(config?: Partial<MemoryGraphConfig>) {
    this.config = {
      endpoint: process.env.MEMORY_GRAPH_ENDPOINT || 'http://localhost:8083',
      apiKey: process.env.MEMORY_GRAPH_API_KEY,
      timeoutMs: parseInt(process.env.MEMORY_GRAPH_TIMEOUT_MS || '30000', 10),
      ...config,
    };
  }

  async connect(): Promise<void> {
    logger.info({ endpoint: this.config.endpoint }, 'Connecting to LLM-Memory-Graph');
    this.connected = true;
    logger.info('Successfully connected to LLM-Memory-Graph');
  }

  async healthCheck(): Promise<AdapterHealth> {
    const start = Date.now();

    if (!this.connected) {
      return unhealthyAdapter('memory_graph', 'Not connected');
    }

    const latencyMs = Date.now() - start;
    return healthyAdapter('memory_graph', latencyMs);
  }

  async disconnect(): Promise<void> {
    logger.info('Disconnecting from LLM-Memory-Graph');
    this.connected = false;
  }

  /**
   * Fetch context lineage
   */
  async fetchContextLineage(query: LineageQuery): Promise<ContextLineage> {
    if (!this.connected) {
      throw new Error('Memory-Graph adapter not connected');
    }

    logger.debug({ query }, 'Fetching context lineage from Memory-Graph');

    // Placeholder implementation
    return {
      lineageId: crypto.randomUUID(),
      rootContextId: query.contextId || 'root',
      createdAt: new Date(),
      depth: 0,
      nodes: [],
      edges: [],
      metadata: {
        totalTokens: 0,
        totalInteractions: 0,
        activeBranches: 0,
        compressionRatio: 1.0,
      },
    };
  }

  /**
   * Fetch interaction graph
   */
  async fetchInteractionGraph(sessionId: string): Promise<InteractionGraph> {
    if (!this.connected) {
      throw new Error('Memory-Graph adapter not connected');
    }

    logger.debug({ sessionId }, 'Fetching interaction graph from Memory-Graph');

    // Placeholder implementation
    return {
      graphId: crypto.randomUUID(),
      sessionId,
      createdAt: new Date(),
      lastUpdated: new Date(),
      statistics: {
        nodeCount: 0,
        edgeCount: 0,
        avgDegree: 0,
        clusteringCoefficient: 0,
        diameter: 0,
        density: 0,
      },
      topics: [],
      entities: [],
    };
  }

  /**
   * Fetch memory snapshot
   */
  async fetchMemorySnapshot(sessionId: string): Promise<MemorySnapshot> {
    if (!this.connected) {
      throw new Error('Memory-Graph adapter not connected');
    }

    logger.debug({ sessionId }, 'Fetching memory snapshot from Memory-Graph');

    // Placeholder implementation
    return {
      snapshotId: crypto.randomUUID(),
      sessionId,
      createdAt: new Date(),
      contextWindowTokens: 0,
      summarizedTokens: 0,
      activeMemories: [],
      retrievalStats: {
        totalRetrievals: 0,
        avgLatencyMs: 0,
        cacheHitRate: 0,
        relevanceAvg: 0,
      },
    };
  }

  /**
   * Fetch graph analytics
   */
  async fetchGraphAnalytics(timeRange: TimeRange): Promise<GraphAnalytics> {
    if (!this.connected) {
      throw new Error('Memory-Graph adapter not connected');
    }

    logger.debug({ timeRange }, 'Fetching graph analytics from Memory-Graph');

    // Placeholder implementation
    return {
      periodStart: timeRange.start,
      periodEnd: timeRange.end,
      totalSessions: 0,
      totalNodesCreated: 0,
      totalEdgesCreated: 0,
      avgSessionDepth: 0,
      avgSessionTokens: 0,
      topTopics: [],
      memoryEfficiency: {
        avgCompressionRatio: 0,
        cacheHitRate: 0,
        retrievalLatencyP50Ms: 0,
        retrievalLatencyP99Ms: 0,
      },
    };
  }
}

File: src/adapters/memory_graph.rs
//! LLM-Memory-Graph Adapter
//!
//! Thin adapter for consuming context lineage and graph-based interaction
//! metadata from LLM-Memory-Graph.
//!
//! This adapter provides read-only access to memory graph data for analytics
//! purposes without modifying any upstream logic.

use super::{AdapterHealth, EcosystemAdapter};
use anyhow::Result;
use async_trait::async_trait;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::atomic::{AtomicBool, Ordering};
use std::time::Instant;
use tracing::{debug, info, instrument};

/// Configuration for Memory-Graph adapter
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MemoryGraphConfig {
    pub endpoint: String,
    pub api_key: Option<String>,
    pub timeout_secs: u64,
}

impl MemoryGraphConfig {
    pub fn from_env() -> Result<Self> {
        Ok(Self {
            endpoint: std::env::var("MEMORY_GRAPH_ENDPOINT")
                .unwrap_or_else(|_| "http://localhost:8083".to_string()),
            api_key: std::env::var("MEMORY_GRAPH_API_KEY").ok(),
            timeout_secs: std::env::var("MEMORY_GRAPH_TIMEOUT_SECS")
                .unwrap_or_else(|_| "30".to_string())
                .parse()
                .unwrap_or(30),
        })
    }
}

/// Context lineage information
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ContextLineage {
    pub lineage_id: String,
    pub root_context_id: String,
    pub created_at: DateTime<Utc>,
    pub depth: u32,
    pub nodes: Vec<LineageNode>,
    pub edges: Vec<LineageEdge>,
    pub metadata: LineageMetadata,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LineageNode {
    pub node_id: String,
    pub node_type: NodeType,
    pub created_at: DateTime<Utc>,
    pub content_hash: String,
    pub token_count: u64,
    pub attributes: HashMap<String, serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum NodeType {
    UserMessage,
    AssistantResponse,
    SystemPrompt,
    ToolCall,
    ToolResult,
    ContextInjection,
    Summary,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LineageEdge {
    pub edge_id: String,
    pub source_node_id: String,
    pub target_node_id: String,
    pub edge_type: EdgeType,
    pub weight: f64,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum EdgeType {
    Follows,
    References,
    Summarizes,
    DerivedFrom,
    ToolInvocation,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LineageMetadata {
    pub total_tokens: u64,
    pub total_interactions: u64,
    pub active_branches: u32,
    pub compression_ratio: f64,
}

/// Graph-based interaction metadata
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InteractionGraph {
    pub graph_id: String,
    pub session_id: String,
    pub created_at: DateTime<Utc>,
    pub last_updated: DateTime<Utc>,
    pub statistics: GraphStatistics,
    pub topics: Vec<TopicCluster>,
    pub entities: Vec<EntityReference>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GraphStatistics {
    pub node_count: u64,
    pub edge_count: u64,
    pub avg_degree: f64,
    pub clustering_coefficient: f64,
    pub diameter: u32,
    pub density: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TopicCluster {
    pub cluster_id: String,
    pub topic: String,
    pub relevance_score: f64,
    pub node_ids: Vec<String>,
    pub keywords: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EntityReference {
    pub entity_id: String,
    pub entity_type: String,
    pub name: String,
    pub first_mentioned: DateTime<Utc>,
    pub mention_count: u32,
    pub node_ids: Vec<String>,
}

/// Memory context snapshot
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MemorySnapshot {
    pub snapshot_id: String,
    pub session_id: String,
    pub created_at: DateTime<Utc>,
    pub context_window_tokens: u64,
    pub summarized_tokens: u64,
    pub active_memories: Vec<ActiveMemory>,
    pub retrieval_stats: RetrievalStats,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ActiveMemory {
    pub memory_id: String,
    pub memory_type: MemoryType,
    pub content_preview: String,
    pub token_count: u64,
    pub relevance_score: f64,
    pub last_accessed: DateTime<Utc>,
    pub access_count: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum MemoryType {
    ShortTerm,
    LongTerm,
    Working,
    Episodic,
    Semantic,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RetrievalStats {
    pub total_retrievals: u64,
    pub avg_latency_ms: f64,
    pub cache_hit_rate: f64,
    pub relevance_avg: f64,
}

/// Query parameters for context lineage
#[derive(Debug, Clone, Default)]
pub struct LineageQuery {
    pub context_id: Option<String>,
    pub session_id: Option<String>,
    pub start_time: Option<DateTime<Utc>>,
    pub end_time: Option<DateTime<Utc>>,
    pub min_depth: Option<u32>,
    pub max_depth: Option<u32>,
    pub include_content: bool,
}

/// LLM-Memory-Graph adapter for consuming graph data
pub struct MemoryGraphAdapter {
    config: MemoryGraphConfig,
    connected: AtomicBool,
}

impl MemoryGraphAdapter {
    pub fn new(config: MemoryGraphConfig) -> Self {
        Self {
            config,
            connected: AtomicBool::new(false),
        }
    }

    /// Fetch context lineage
    #[instrument(skip(self))]
    pub async fn fetch_context_lineage(&self, query: LineageQuery) -> Result<ContextLineage> {
        if !self.connected.load(Ordering::Relaxed) {
            anyhow::bail!("Memory-Graph adapter not connected");
        }

        debug!(?query.context_id, "Fetching context lineage from Memory-Graph");

        // Placeholder implementation
        Ok(ContextLineage {
            lineage_id: uuid::Uuid::new_v4().to_string(),
            root_context_id: query.context_id.unwrap_or_else(|| "root".to_string()),
            created_at: Utc::now(),
            depth: 0,
            nodes: Vec::new(),
            edges: Vec::new(),
            metadata: LineageMetadata {
                total_tokens: 0,
                total_interactions: 0,
                active_branches: 0,
                compression_ratio: 1.0,
            },
        })
    }

    /// Fetch interaction graph for a session
    #[instrument(skip(self))]
    pub async fn fetch_interaction_graph(&self, session_id: &str) -> Result<InteractionGraph> {
        if !self.connected.load(Ordering::Relaxed) {
            anyhow::bail!("Memory-Graph adapter not connected");
        }

        debug!(session_id = %session_id, "Fetching interaction graph from Memory-Graph");

        // Placeholder implementation
        Ok(InteractionGraph {
            graph_id: uuid::Uuid::new_v4().to_string(),
            session_id: session_id.to_string(),
            created_at: Utc::now(),
            last_updated: Utc::now(),
            statistics: GraphStatistics {
                node_count: 0,
                edge_count: 0,
                avg_degree: 0.0,
                clustering_coefficient: 0.0,
                diameter: 0,
                density: 0.0,
            },
            topics: Vec::new(),
            entities: Vec::new(),
        })
    }

    /// Fetch memory snapshot for a session
    #[instrument(skip(self))]
    pub async fn fetch_memory_snapshot(&self, session_id: &str) -> Result<MemorySnapshot> {
        if !self.connected.load(Ordering::Relaxed) {
            anyhow::bail!("Memory-Graph adapter not connected");
        }

        debug!(session_id = %session_id, "Fetching memory snapshot from Memory-Graph");

        // Placeholder implementation
        Ok(MemorySnapshot {
            snapshot_id: uuid::Uuid::new_v4().to_string(),
            session_id: session_id.to_string(),
            created_at: Utc::now(),
            context_window_tokens: 0,
            summarized_tokens: 0,
            active_memories: Vec::new(),
            retrieval_stats: RetrievalStats {
                total_retrievals: 0,
                avg_latency_ms: 0.0,
                cache_hit_rate: 0.0,
                relevance_avg: 0.0,
            },
        })
    }

    /// Get graph statistics for analytics
    #[instrument(skip(self))]
    pub async fn fetch_graph_analytics(
        &self,
        start: DateTime<Utc>,
        end: DateTime<Utc>,
    ) -> Result<GraphAnalytics> {
        if !self.connected.load(Ordering::Relaxed) {
            anyhow::bail!("Memory-Graph adapter not connected");
        }

        debug!("Fetching graph analytics from Memory-Graph");

        // Placeholder implementation
        Ok(GraphAnalytics {
            period_start: start,
            period_end: end,
            total_sessions: 0,
            total_nodes_created: 0,
            total_edges_created: 0,
            avg_session_depth: 0.0,
            avg_session_tokens: 0.0,
            top_topics: Vec::new(),
            memory_efficiency: MemoryEfficiency {
                avg_compression_ratio: 0.0,
                cache_hit_rate: 0.0,
                retrieval_latency_p50_ms: 0.0,
                retrieval_latency_p99_ms: 0.0,
            },
        })
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GraphAnalytics {
    pub period_start: DateTime<Utc>,
    pub period_end: DateTime<Utc>,
    pub total_sessions: u64,
    pub total_nodes_created: u64,
    pub total_edges_created: u64,
    pub avg_session_depth: f64,
    pub avg_session_tokens: f64,
    pub top_topics: Vec<String>,
    pub memory_efficiency: MemoryEfficiency,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MemoryEfficiency {
    pub avg_compression_ratio: f64,
    pub cache_hit_rate: f64,
    pub retrieval_latency_p50_ms: f64,
    pub retrieval_latency_p99_ms: f64,
}

#[async_trait]
impl EcosystemAdapter for MemoryGraphAdapter {
    #[instrument(skip(self))]
    async fn connect(&self) -> Result<()> {
        info!(endpoint = %self.config.endpoint, "Connecting to LLM-Memory-Graph");

        self.connected.store(true, Ordering::Relaxed);

        info!("Successfully connected to LLM-Memory-Graph");
        Ok(())
    }

    async fn health_check(&self) -> Result<AdapterHealth> {
        let start = Instant::now();

        if !self.connected.load(Ordering::Relaxed) {
            return Ok(AdapterHealth::unhealthy("memory_graph", "Not connected"));
        }

        let latency_ms = start.elapsed().as_millis() as u64;
        Ok(AdapterHealth::healthy("memory_graph", latency_ms))
    }

    #[instrument(skip(self))]
    async fn disconnect(&self) -> Result<()> {
        info!("Disconnecting from LLM-Memory-Graph");
        self.connected.store(false, Ordering::Relaxed);
        Ok(())
    }
}
